File: midonet-cluster/src/main/java/org/midonet/cluster/rest_api/neutron/models/QOSPolicy.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.cluster.rest_api.neutron.models;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.MoreObjects;
import org.apache.commons.collections4.ListUtils;
import org.midonet.cluster.data.*;
import org.midonet.cluster.models.Neutron;
import org.midonet.util.collection.ListUtil;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

@ZoomClass(clazz = Neutron.NeutronQOSPolicy.class)
public class QOSPolicy extends ZoomObject {

    public QOSPolicy() {}

    public QOSPolicy(UUID qosId) {
        this.id = qosId;
    }

    @ZoomField(name = "id")
    public UUID id;

    @JsonProperty("bandwidth_limit_rules")
    @ZoomField(name = "bandwidth_limit_rules")
    public List<QOSRuleBWLimit>  bandwidthLimitRules;

    @JsonProperty("bandwidth_limit_rules")
    @ZoomField(name = "dscp_marking_rules")
    public List<QOSRuleDSCP> dscpMarkingRules;

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        QOSPolicy that = (QOSPolicy) o;

        return Objects.equals(id, that.id) &&
                ListUtils.isEqualList(bandwidthLimitRules,
                        that.bandwidthLimitRules) &&
                ListUtils.isEqualList(dscpMarkingRules,
                        that.dscpMarkingRules);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id,
                ListUtils.hashCodeForList(bandwidthLimitRules),
                ListUtils.hashCodeForList(dscpMarkingRules));
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
            .omitNullValues()
            .add("id", id)
            .add("bandwidthLimitRules", ListUtil.toString(bandwidthLimitRules))
            .add("dscpMarkingRules", ListUtil.toString(dscpMarkingRules))
            .toString();
    }
}

File: tests/mdts/lib/qos_policy.py
# Copyright 2016 Midokura SARL
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from mdts.lib.resource_base import ResourceBase

class QOSPolicy(ResourceBase):

    def __init__(self, api, context, data):
        super(QOSPolicy, self).__init__(api, context, data)

    def build(self):
        tenant_id = self._get_tenant_id()
        self._mn_resource = self._api.add_qos_policy()
        self._mn_resource.tenant_id(tenant_id)
        self._mn_resource.name(self._get_name())
        self._mn_resource.description(self._data.get('description'))

        self._mn_resource.create()

        for bw_rule in self._data.get('bw_limit_rules') or []:
            self.add_bw_limit_rule(bw_rule['bw__limit_rule'])

        for dscp_rule in self._data.get('dscp_rules') or []:
            self.add_dscp_rule(dscp_rule['dscp_rule'])

    def description(self, desc):
        self._data["description"] = desc
        self._mn_resource.description(desc).update()

    def shared(self, shr):
        self._data["shared"] = shr
        self._mn_resource.shared(shr).update()

    def destroy(self):
        self._mn_resource.delete()

    def get_id(self):
        return self._mn_resource.get_id()

    def add_bw_limit_rule(self, rule):
        qos_rule = self._mn_resource.add_bw_limit_rule()
        qos_rule.max_kbps(rule['max_kbps'])
        qos_rule.max_burst_kbps(rule['max_burst_kbps'])
        qos_rule.create()
        return qos_rule

    def get_bw_limit_rules(self):
        return self._mn_resource.get_bw_limit_rules()

    def clear_bw_limit_rules(self):
        for rule in self.get_bw_limit_rules():
            rule.delete()

    def add_dscp_rule(self, rule):
        qos_rule = self._mn_resource.add_dscp_rule()
        qos_rule.dscp_mark(rule['dscp_mark'])
        qos_rule.create()
        return qos_rule

    def get_dscp_rules(self):
        return self._mn_resource.get_dscp_rules()

    def clear_dscp_rules(self):
        for rule in self.get_dscp_rules():
            rule.delete()
